### PR TITLE
Maria Velez - Octos - OO-RideShare

### PR DESCRIPTION
**Pull request by velezdev**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I used a 'change status' method to toggle between driver being available and non-available. Isolating this behavior onto a method on its own made calling the method much easier than incorporating this behavior inside some other method.
Describe any examples of composition that you encountered in this project, if any | Anytime I called a method in another class, I was using composition. 
Describe a _nominal test_ that you wrote for this assignment. | In trip_spec.rb, I wrote a test called "it "tests duration of trip in seconds between end and start times" and it does exactly as the title describes.
Describe an _edge case test_ that you wrote for this assignment. | For the same method as described above ('duration'), I wrote a test called 'it "calculates duration when end time is nil" do'.
Describe a concept that you gained more clarity on as you worked on this assignment. | Calling method names outside of its class and also recognizing that the variables initialized and listed in attr_reader had inherent method-like behaviors.

### CODE REVIEW

**Comment by 0velez0** [comment]
The project is not complete and there is a lot of room for improvement in the testing, as well as making sure the current tests pass. Turning in this assignment on the heels of starting the 'Hotel' project makes me want to continue moving forward and not spend more time on it.

File: lib/driver.rb [hunk 1]
```diff
@@ -19,29 +23,71 @@ def initialize(input)
...
+
+    def average_revenue
+      (total_revenue/finished_trips.length).round(2)
     end
-  end
-end
+
+    def finished_trips
+      trips_reject{|trip| trip.end_time == nil}
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:90]: I love that you've broken this out as a separate helper method! This makes the other code in this file much easier to read.

File: lib/trip.rb [hunk 2]
```diff
@@ -13,9 +15,41 @@ def initialize(input)
...
+      @end_time = Time.new
+    end # ends "finish_trip!" method
+
+    def finished?
+      return end_time != nil
+    end # ends "finished?" method
+
+    def finished_trips
+      trips.select {|trip| trip.end_time == nil}
     end
```
Comment (droberts-sea) [inline lib/trip.rb:44]: I don't know that it makes sense to include a `finished_trips` instance method on `Trip`. Bringing it back to Metz, that's like asking a trip "hey, which of your trips are finished?" The trip would probably respond "I _am_ a trip, I don't _have_ trips".

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -15,6 +16,43 @@ def initialize
...
+      # Your code should automatically assign a driver to the trip
+      # For this initial version, choose the first driver whose status is :AVAILABLE
+      # The trip hasn't finished yet!
+      available_driver = drivers.find {|driver| driver.status == :AVAILABLE }
+
+      # Added this line due to an error in this test: it "raise an exception if you try to request a trip when drivers are UNAVAILABLE" do
+      raise "No available drivers" if available_driver.nil?
+
+      selected_passenger = passengers.find {|passenger| passenger.id == passenger_id}
+      max_trip_id = trips.map {|trip| trip.id}.max
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:30]: On line 29, why not call the existing `find_passenger` method?

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -75,6 +75,111 @@
...
+    #   trip = RideShare::Trip.new(trip_data1)
+    #   @driver.add_trip(trip1)
+    #   trip = RideShare::Trip.new(trip_data2)
+    #   @driver.add_trip(trip2)
+    #
+    # end # ends "it "checks average_revenue method" do"
+
+    it "calculates the driver's average_revenue per hour spent driving (using the above formula for revenue)" do
+      @driver.average_revenue.must_equal 1010101010
+    end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:143]: This number does not seem right to me.

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -75,6 +75,111 @@
     it "returns zero if no trips" do
       driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
       driver.average_rating.must_equal 0
+    end # ends 'it "returns zero if no trips" do'
+  end # ends 'describe "average_rating method" do'
+
+
+  describe "total_revenue method" do
+    it "calculates the total revenue" do
+      trips = [
```
Comment (droberts-sea) [inline specs/driver_spec.rb:84]: Other questions that I would like to see these tests ask, for all the wave 1 methods (on both `Driver` and `Passenger`):

- What happens if the driver or passenger has no trips?
- What happens if the driver or passenger has an incomplete trip?

File: specs/trip_dispatch_spec.rb [hunk 6]
```diff
@@ -12,12 +13,54 @@
...
+      # # available_driver = @drivers.find_driver(2)
+      # available_driver.name.must_equal "Emory Rosenbaum"
+      # # available_driver.id.must_equal 2
+      # available_driver.status.must_equal :UNAVAILABLE
+    end
+
+    it "raise an exception if you try to request a trip when drivers are UNAVAILABLE" do
+      dispatcher = RideShare::TripDispatcher.new
+      dispatcher.drivers.each do |driver|
+        driver.status = :UNAVAILABLE
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:58]: Good work catching this interesting edge case.

File: lib/driver.rb (hunk 1 above)
Comment (droberts-sea) [reply lib/driver.rb:90]: One question: why not call the `trip.finished?` method you defined in `trip.rb`?

This is exactly the sort of thing Metz is talking about when she says depend on behavior not data. Calling the `finished?` method is _behavior_, whereas checking `trip.end_time == nil` is looking at _data_.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Your commit messages look good, but I would like to see more frequent commits. If you're having trouble remembering to commit regularly, [you might find this article interesting](https://dev.to/rpalo/plan-your-commits)
Answer comprehension questions | yes - Your answer on composition gets at this, but I want to call out this vocab explicitly. By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | no - looks like these were missed
`Driver` has a method to calculate total revenue | yes
Complete tests | some missing - see inline
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | yes
Complete tests | some missing
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | N/A
Appropriate use of Ruby's `Date` | N/A
Appropriate helper methods were made to help with complex logic | N/A
Complete tests | N/A
**Overall** |

This is a good start. Your code is for the most part well-organized and easy to read, and I like the way you've demonstrated functional decomposition.

The big place where I see room for improvement is around test coverage. You're doing a good job of identifying nominal cases, but there are many interesting edge cases you're not hitting. You should work on finding the "edge" between working and not working, which means identifying the spots where a method you've written might do something funny. I've tried to outline some of these in the comments below.

I'm also a little concerned about the speed at which you're able to turn ideas into code. I don't have direct evidence of this, but the missing methods on `Passenger` from wave 1, and the number of test cases you have stubbed out but not implemented, indicate to me that you're struggling to keep a cadence.

Here are a few things you can try when you find yourself staring at one of the requirements with no idea where to start:
- Re-read the requirement. Paraphrase it to yourself, write it down with pencil and paper, make sure you understand what's being asked.
- Solve the problem by hand. What if you had a piece of paper with this information instead of a computer, how would you come up with the answer?
- Break off the smallest possible thing you can think of. "I'm calculating the passenger's total cost, so I guess I'm going to need to get the cost from a trip. How do I do that? Well, if the trip is stored in an instance variable called `trip`, then it would be `trip.cost`." Then you can build backward from that: "OK, now how do I get a trip into an instance variable?"

Hopefully that's helpful. I've seen you putting in the hours and I really appreciate that dedication - keep up the hard work.

**Pull request closed without merging**